src/include/falconn/lsh_nn_table.h

```diff
@@ -88,36 +79,158 @@ class LSHNearestNeighborTable {
+
+  ///
+  /// Returns the keys corresponding to candidates in the probing sequence for q
+  /// that have distance at most threshold.
+  ///
+  virtual void find_near_neighbors(
+      const PointType& q,
+      typename PointTypeTraits<PointType>::ScalarType threshold,
+      std::vector<KeyType>* result) = 0;
 
   ///
   /// Returns the keys of all candidates in the probing sequence for q.
-  /// Every candidate key occurs only once in the result. The
-  /// candidates are returned in the order of their first occurrence in the
-  /// probing sequence.
+  /// See the documentation for LSHNearestNeighborQuery.
   ///
   virtual void get_unique_candidates(const PointType& q,
                                      std::vector<KeyType>* result) = 0;
 
   ///
+  /// Returns the multiset of all candidate keys in the probing sequence for q.
+  /// See the documentation for LSHNearestNeighborQuery.
+  ///
+  virtual void get_candidates_with_duplicates(const PointType& q,
+                                              std::vector<KeyType>* result) = 0;
+  
+  ///
   /// Resets the query statistics.
   ///
   virtual void reset_query_statistics() = 0;
 
   ///
   /// Returns the current query statistics.
-  ///
-  /// TODO: figure out the right semantics here: should the average distance
-  /// time be averaged over all queries or only the near(est) neighbor queries?
+  /// See the documentation for LSHNearestNeighborQuery.
   ///
   virtual QueryStatistics get_query_statistics() = 0;
 
+
+  virtual ~LSHNearestNeighborQueryPool() {}
+};
+
+///
+/// Common interface shared by all LSH table wrappers.
+///
+/// The template parameter PointType should be one of the two point types
+/// introduced above (DenseVector and SparseVector), e.g., DenseVector<float>.
+///
+/// The KeyType template parameter is optional and the default int32_t is
+/// sufficient for up to 10^9 points.
+///
+template <typename PointType, typename KeyType = int32_t>
+class LSHNearestNeighborTable {
+ public:
+  ///
+  /// A special constant for set_max_num_candidates which is effectively
+  /// equivalent to infinity.
+  ///
+  static const int_fast64_t kNoMaxNumCandidates = -1;
+ 
+  ///
+  /// This function constructs a new query object. The query object holds
+  /// all per-query state and executes table lookups.
+  ///
+  virtual std::unique_ptr<LSHNearestNeighborQuery<PointType, KeyType>>
+  construct_query_object(int_fast64_t num_probes = -1,
+                         int_fast64_t max_num_candidates = -1) const = 0;
+  
+  ///
+  /// This function constructs a new query pool. The query pool holds
+  /// a set of query objects and supports an interface that can be safely
+  /// called from multiple threads.
+  ///
+  virtual std::unique_ptr<LSHNearestNeighborQueryPool<PointType, KeyType>>
+  construct_query_pool(int_fast64_t num_probes = -1,
+                       int_fast64_t max_num_candidates = -1,
+                       int_fast64_t num_query_objects = -1) const = 0;
```

Comment:
It uses a small constant factor times the number of available hardware threads. This should be sufficient for use in a good thread pool implementation (that has number of parallel threads = number of hardware threads).